Add the release note for TiDB v5.1 (#5840)
pingcap/docs: releases/release-5.1.0.md

| [`tidb_partition_prune_mode`](/system-variables.md#tidb_partition_prune_mode-new-in-v51) | Newly added | Specifies whether to enable dynamic mode for partitioned tables. This feature is experimental. The default value of this variable is `static`, which means the dynamic mode for partitioned tables is disabled by default.  |

### Configuration file parameters

| Configuration file | Configuration item | Change type   | Description   |
|:----------|:-----------|:-----------|:-----------|
| TiDB configuration file | [`security.enable-sem`](tidb-configuration-file.md#enable-sem)  | Newly added  | Controls whether to enable the Security Enhanced Mode (SEM). The default value of this configuration item is `false`, which means the SEM is disabled. |
| TiDB configuration file | [`performance.committer-concurrency`](/tidb-configuration-file.md#committer-concurrency)  | Modified  | The number of goroutines for requests related to executing commits in the commit phase of a single transaction. The default value is changed from `16` to `128`. |

> | TiDB configuration file | [`performance.tcp-no-delay`](/tidb-configuration-file.md#tcp-no-delay)  | Newly added  | Determines whether to enable TCP_NODELAY at the TCP layer. The default value is `true`, which means TCP_NODELAY is enabled. |

Author: ```suggestion
| TiDB configuration file | [`performance.tcp-no-delay`](/tidb-configuration-file.md#tcp-no-delay)  | Newly added  | Determines whether to enable TCP_NODELAY at the TCP layer. The default value is `true`, which means TCP_NODELAY is enabled. |
| TiDB configuration file | [`performance.enforce-mpp`](/tidb-configuration-file.md#enforce-mpp)  | Newly added  | Controls whether TiDB ignores cost estimates of Optimizer at the instance level and enforces the MPP mode. The default value is `false`. |

```

| TiKV configuration file | [`resolved-ts.advance-ts-interval`](/tikv-configuration-file.md#advance-ts-interval)  | Newly added  | The interval at which the `resolved-ts` is forwarded. The default value is `"1s"`. You can change the value dynamically. |
| TiKV configuration file | [`resolved-ts.scan-lock-pool-size`](/tikv-configuration-file.md#scan-lock-pool-size)  | Newly added  | The number of threads that TiKV uses to scan the MVCC (multi-version concurrency control) lock data when initializing the `resolved-ts`. The default value is `2`. |

### Others

- Upgrade the Go compiler version of TiDB from go1.13.7 to go1.16.4, which improves the TiDB performance. If you are a TiDB developer, upgrade your Go compiler version to ensure a smooth compilation. 
- Avoid creating tables with clustered indexes in the cluster that uses TiDB Binlog during the TiDB rolling upgrade.
- Avoid executing statements like `alter table ... modify column` or `alter table ... change column` during the TiDB rolling upgrade.

> - Since v5.1, setting the replica of system tables, when building TiFlash replicas for each table, is no longer supported. Before upgrading the cluster, you need to clear the relevant system table replicas; otherwise, the upgrade will fail..

Reviewer: ```suggestion
- Since v5.1, setting the replica of system tables, when building TiFlash replicas for each table, is no longer supported. Before upgrading the cluster, you need to clear the relevant system table replicas; otherwise, the upgrade will fail.
```

    The supported dynamic privileges are as follows:

    - `BACKUP_ADMIN`
    - `RESTORE_ADMIN`
    - `ROLE_ADMIN`
    - `CONNECTION_ADMIN`
    - `SYSTEM_VARIABLES_ADMIN`

    You can also use plugins to add new privileges. To check out all supported privileges, execute the `SHOW PRIVILEGES` statement. [User document](/privilege-management.md)

- Add a new configuration item for the Security Enhanced Mode (SEM), which divides the TiDB administrator privileges in a finer-grained way. 

    The Security Enhanced Mode is disabled by default. To enable it, see the [user document](/system-variables.md#tidb_enable_enhanced_security).

- Enhance the capability of changing column types online. Support changing the column type online using the `ALTER TABLE` statement, including but not limited to:

    - Changing `VARCHAR` to `BIGINT`
    - Modifying the `DECIMAL` precision
    - Compressing the length of `VARCHAR(10)` to `VARCHAR(5)`

    [User document](/sql-statements/sql-statement-modify-column.md)

> - Introduce a new SQL syntax `AS OF TIMESTAMP` to perform Stale Read, a new feature used to read historical data from a specified point in time or from a specified time range.

Author: ```suggestion
- Introduce a new SQL syntax `AS OF TIMESTAMP` to perform Stale Read, a new experimental feature used to read historical data from a specified point in time or from a specified time range.
```

- Introduce a new statistics type `tidb_analyze_version = 2` (experimental feature).

    `tidb_analyze_version` is set to `2` by default, which avoids the large errors that might occur in the large data volume caused by hash conflicts in Version 1 and maintains the estimation accuracy in most scenarios.

    [User document](/statistics.md)

### Transaction

+ Support the Lock View feature (Experimental Feature)

    The Lock View feature provides more information about lock conflicts and lock waits of pessimistic locks, which helps DBAs to observe transaction locking conditions and troubleshoot deadlock problems. [#24199](https://github.com/pingcap/tidb/issues/24199)

    User document:
        - View the pessimistic locks and other locks that currently occur on all TiKV nodes in the clusters: 
[`DATA_LOCK_WAITS`](/information-schema/information-schema-data-lock-waits.md)
        - View several deadlock errors that recently occured on the TiDB nodes:
[`DEADLOCKS`](/information-schema/information-schema-deadlocks.md)
        - View the transaction information executed currently on the TiDB nodes:
[`TIDB_TRX`](/information-schema/information-schema-tidb-trx.md)

### Performance

> + Stale read of data replicas

Author: ```suggestion
+ Stale read of data replicas (Experimental)
```

    - Solve multiple scheduling issues in the scaling process

        - Optimize the generation process of replica snapshots to solve slow scheduling issues during scaling: 
[#3563](https://github.com/tikv/pd/issues/3563) [#10059](https://github.com/tikv/tikv/pull/10059) [#10001](https://github.com/tikv/tikv/pull/10001)
        - Solve slow scheduling issues caused by heartbeat pressure due to traffic changes 
[#3693](https://github.com/tikv/pd/issues/3693) [#3739](https://github.com/tikv/pd/issues/3739) [#3728](https://github.com/tikv/pd/issues/3728) [#3751](https://github.com/tikv/pd/issues/3751)
        - Reduce the space discrepancies of large clusters due to scheduling, and optimize the scheduling formula to prevent the bursting issue (which is similar to heterogeneous space clusters) caused by large compression rate discrepancies [#3592](https://github.com/tikv/pd/issues/3592) [#10005](https://github.com/tikv/tikv/pull/10005)

+ Tools

    + Backup & Restore (BR)

        - Support backing up and restoring system tables in the `mysql` schema [#1143](https://github.com/pingcap/br/pull/1143) [#1078](https://github.com/pingcap/br/pull/1078)

>         - Support S3-compatible storage (based on the virtual-host addressing mode) [#10243](https://github.com/tikv/tikv/pull/10243)

Author: ```suggestion
        - Support the S3-compatible storage that is based on the virtual-host addressing mode [#10243](https://github.com/tikv/tikv/pull/10243)
```

| [`tidb_enable_enhanced_security`](/system-variables.md#tidb_enable_enhanced_security) | Newly added | Indicates whether the TiDB server you are connected to has the Security Enhanced Mode (SEM) enabled. This variable setting cannot be changed without restarting the TiDB server. |
| [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-new-in-v51) | Newly added | Controls whether to ignore the optimizer's cost estimation and to forcibly use the MPP mode for query execution. The data type of this variable is `BOOL` and the default value is `false`. |
| [`tidb_partition_prune_mode`](/system-variables.md#tidb_partition_prune_mode-new-in-v51) | Newly added | Specifies whether to enable dynamic prune mode for partitioned tables. This feature is experimental. The default value of this variable is `static`, which means the dynamic mode for partitioned tables is disabled by default.  |

### Configuration file parameters

| Configuration file | Configuration item | Change type   | Description   |
|:----------|:-----------|:-----------|:-----------|

> | TiDB configuration file | [`security.enable-sem`](tidb-configuration-file.md#enable-sem)  | Newly added  | Controls whether to enable the Security Enhanced Mode (SEM). The default value of this configuration item is `false`, which means the SEM is disabled. |

Author: ```suggestion
| TiDB configuration file | [`security.enable-sem`](/tidb-configuration-file.md#enable-sem)  | Newly added  | Controls whether to enable the Security Enhanced Mode (SEM). The default value of this configuration item is `false`, which means the SEM is disabled. |
```

| TiDB configuration file | [`performance.committer-concurrency`](/tidb-configuration-file.md#committer-concurrency)  | Modified  | Controls the concurrency number for requests related to commit operations in the commit phase of a single transaction. The default value is changed from `16` to `128`. |
| TiDB configuration file | [`performance.tcp-no-delay`](/tidb-configuration-file.md#tcp-no-delay)  | Newly added  | Determines whether to enable TCP_NODELAY at the TCP layer. The default value is `true`, which means TCP_NODELAY is enabled. |

> | TiDB configuration file | [`performance.enforce-mpp`](/tidb-configuration-file.md#enforce-mpp)  | Newly added  | Controls whether TiDB ignores cost estimates of Optimizer at the instance level and enforces the MPP mode. The default value is `false`. |

Reviewer 2: suggest add: This configuration item is the initial value of the system variable `tidb_enforce_mpp`.

> | TiDB configuration file | [`pessimistic-txn.deadlock-history-capacity`](tidb-configuration-file.md#deadlock-history-capacity)  | Newly added  | Sets the maximum number of deadlock events that can be recorded in the [`INFORMATION_SCHEMA.DEADLOCKS`](/information-schema/information-schema-deadlocks.md) table of a single TiDB server. The default value is `10`. |

Author: ```suggestion
| TiDB configuration file | [`pessimistic-txn.deadlock-history-capacity`](/tidb-configuration-file.md#deadlock-history-capacity)  | Newly added  | Sets the maximum number of deadlock events that can be recorded in the [`INFORMATION_SCHEMA.DEADLOCKS`](/information-schema/information-schema-deadlocks.md) table of a single TiDB server. The default value is `10`. |
```

    - Optimize lock operations to avoid that running DDL statements and read operations are blocked by each other
    - Optimize cleanup of expired data by TiFlash 
    - Support further filtering of query filters on `timestamp` columns at the TiFlash storage level
    - Improve the startup and scalability speed of TiFlash when a large number of tables are in a cluster
    - Improve TiFlash compatibility when running on unknown CPUs

+ PD
    - Avoid unexpected statistics after adding the `scatter region` scheduler [#3602](https://github.com/pingcap/pd/pull/3602)
    - Solve multiple scheduling issues in the scaling process

        - Optimize the generation process of replica snapshots to solve slow scheduling issues during scaling: 
[#3563](https://github.com/tikv/pd/issues/3563) [#10059](https://github.com/tikv/tikv/pull/10059) [#10001](https://github.com/tikv/tikv/pull/10001)
        - Solve slow scheduling issues caused by heartbeat pressure due to traffic changes 
[#3693](https://github.com/tikv/pd/issues/3693) [#3739](https://github.com/tikv/pd/issues/3739) [#3728] (https://github.com/tikv/pd/issues/3728) [#3751](https://github.com/tikv/pd/issues/3751)

>         - Reduce the space discrepancies of large clusters due to scheduling, and optimize the scheduling formula to prevent the bursting issue (which is similar to heterogeneous space clusters) caused by large compression rate discrepancies    [#3592](https://github.com/tikv/pd/issues/3592) [#10005](https://github.com) /tikv/tikv/pull/10005)

Reviewer 3: ```suggestion
        - Reduce the space discrepancies of large clusters due to scheduling, and optimize the scheduling formula to prevent the bursting issue (which is similar to heterogeneous space clusters) caused by large compression rate discrepancies    [#3592](https://github.com/tikv/pd/issues/3592) [#10005](https://github.com/tikv/tikv/pull/10005)
```

+ Tools

    + Backup & Restore (BR)

>         - Support backing up and restoring system tables in the `mysql` schema [#1143](https://github.com/pingcap/br/pull/1143) [#1078](https://github.com/pingcap/br/pull/1078)

Reviewer 3: ```suggestion
        - Support backing up and restoring system tables in the `mysql` schema [#1143](https://github.com/pingcap/br/pull/1143) [#872](https://github.com/pingcap/br/issues/872) [#679](https://github.com/pingcap/br/issues/679)
```

>         - Support S3-compatible storage (based on the virtual-host addressing mode) [#10243](https://github.com/tikv/tikv/pull/10243)

Reviewer 3: ```suggestion
        - Support S3-compatible storage (based on the virtual-host addressing mode) [#484](https://github.com/pingcap/br/issues/484))
```

| Configuration file | Configuration item | Change type   | Description   |
|:----------|:-----------|:-----------|:-----------|
| TiDB configuration file | [`security.enable-sem`](tidb-configuration-file.md#enable-sem)  | Newly added  | Controls whether to enable the Security Enhanced Mode (SEM). The default value of this configuration item is `false`, which means the SEM is disabled. |
| TiDB configuration file | [`performance.committer-concurrency`](/tidb-configuration-file.md#committer-concurrency)  | Modified  | The number of goroutines for requests related to executing commits in the commit phase of a single transaction. The default value is changed from `16` to `128`. |
| TiDB configuration file | [`performance.tcp-no-delay`](/tidb-configuration-file.md#tcp-no-delay)  | Newly added  | Determines whether to enable TCP_NODELAY at the TCP layer. The default value is `true`, which means TCP_NODELAY is enabled. |
| TiDB configuration file | [`performance.enforce-mpp`](/tidb-configuration-file.md#enforce-mpp)  | Newly added  | Controls whether TiDB ignores cost estimates of Optimizer at the instance level and enforces the MPP mode. The default value is `false`. |

> 

Author: ```suggestion
```

    - Improve TiFlash compatibility when running on unknown CPUs

+ PD
    - Avoid unexpected statistics after adding the `scatter region` scheduler [#3602](https://github.com/pingcap/pd/pull/3602)
    - Solve multiple scheduling issues in the scaling process

        - Optimize the generation process of replica snapshots to solve slow scheduling issues during scaling: 
[#3563](https://github.com/tikv/pd/issues/3563) [#10059](https://github.com/tikv/tikv/pull/10059) [#10001](https://github.com/tikv/tikv/pull/10001)
        - Solve slow scheduling issues caused by heartbeat pressure due to traffic changes 
[#3693](https://github.com/tikv/pd/issues/3693) [#3739](https://github.com/tikv/pd/issues/3739) [#3728](https://github.com/tikv/pd/issues/3728) [#3751](https://github.com/tikv/pd/issues/3751)
        - Reduce the space discrepancies of large clusters due to scheduling, and optimize the scheduling formula to prevent the bursting issue (which is similar to heterogeneous space clusters) caused by large compression rate discrepancies [#3592](https://github.com/tikv/pd/issues/3592) [#10005](https://github.com/tikv/tikv/pull/10005)

> 

Reviewer 3: ```suggestion
```

+ Tools

    + Backup & Restore (BR)

        - Support backing up and restoring system tables in the `mysql` schema [#1143](https://github.com/pingcap/br/pull/1143) [#1078](https://github.com/pingcap/br/pull/1078)
        - Support S3-compatible storage (based on the virtual-host addressing mode) [#10243](https://github.com/tikv/tikv/pull/10243)
        - Optimize the format of backupmeta to reduce memory usage [#1171](https://github.com/pingcap/br/pull/1171)

> 

Reviewer 3: ```suggestion
```

    + TiCDC

        - Improve the descriptions of some log messages to be clearer and more useful for diagnosing problems [#1759](https://github.com/pingcap/ticdc/pull/1759)
        - Support the back pressure feature to allow the TiCDC scanning speed to sense the downstream processing capacity [#10151](https://github.com/tikv/tikv/pull/10151)
        - Reduce memory usage when TiCDC performs the initial scan  [#10133](https://github.com/tikv/tikv/pull/10133)
        - Improve the cache hit rate for the TiCDC Old Value in pessimistic transactions [#10089](https://github.com/tikv/tikv/pull/10089)

> 

Reviewer 3: ```suggestion
```

    + Dumpling

        - Improve the logic of exporting data from TiDB v4.0 to avoid TiDB becoming out of memory (OOM) [#273](https://github.com/pingcap/dumpling/pull/273)

        - Fix the issue that no error is output when a backup operation fails [#280](https://github.com/pingcap/dumpling/pull/280)

    + TiDB Lightning

        - Improve data importing speed. The optimization results show that the speed of importing TPC-C data is increased by 30%, and the speed of importing large tables (2TB+) with more indexes (5 indexes) is increased by more than 50%. [#753](https://github.com/pingcap/br/pull/753)
        - Add a pre-check on the data to be imported and also on the target cluster before importing, and report errors to reject the import process if it does not meet the import requirements [#999](https://github.com/pingcap/br/pull/999)
        - Optimize the timing of checkpoint updates on the Local backend to improve performance of restarting from breakpoints[#1080](https://github.com/pingcap/br/pull/1080)

> 
> 

Reviewer 3: ```suggestion
```
